model/waterfall.go

```diff
@@ -0,0 +1,168 @@
+package model
+
+import (
+	"context"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/build"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+const (
+	DefaultWaterfallVersionLimit = 5
+	MaxWaterfallVersionLimit     = 300
+)
+
+type WaterfallTask struct {
+	Id          string `bson:"_id" json:"_id"`
+	DisplayName string `bson:"display_name" json:"display_name,omitempty"`
+	Status      string `bson:"status" json:"status,omitempty"`
+}
+
+type WaterfallBuild struct {
+	Id          string          `bson:"_id" json:"_id"`
+	Activated   bool            `bson:"activated" json:"activated,omitempty"`
+	DisplayName string          `bson:"display_name" json:"display_name,omitempty"`
+	Version     string          `bson:"version" json:"version,omitempty"`
+	Tasks       []WaterfallTask `bson:"tasks" json:"tasks,omitempty"`
+}
+
+type WaterfallBuildVariant struct {
+	Id          string           `bson:"_id" json:"_id"`
+	DisplayName string           `bson:"display_name" json:"display_name,omitempty"`
+	Builds      []WaterfallBuild `bson:"builds" json:"builds,omitempty"`
+}
+
+type WaterfallOptions struct {
+	Limit      int
+	Requesters []string
+}
+
+// GetWaterfallVersions returns `opts.limit` versions for a given project.
+// TODO: It will eventually return `opts.limit` activated versions that also satisfy any given filters.
+func GetWaterfallVersions(ctx context.Context, projectId string, opts WaterfallOptions) ([]Version, error) {
+	invalidRequesters, _ := utility.StringSliceSymmetricDifference(opts.Requesters, evergreen.SystemVersionRequesterTypes)
+	if len(invalidRequesters) > 0 {
+		return nil, errors.Errorf("invalid requesters '%s'", invalidRequesters)
+	}
+	match := bson.M{
+		VersionIdentifierKey: projectId,
+		VersionRequesterKey: bson.M{
+			"$in": opts.Requesters,
+		},
+	}
+
+	// TODO DEVPROD-10177: Add revision order logic to handle pagination.
+
+	pipeline := []bson.M{{"$match": match}}
+	pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
+	limit := defaultVersionLimit
```

Comment:
Is there a reason this would be `defaultVersionLimit` instead of the `DefaultWaterfallVersionLimit`? What's the difference between the two constants?

Reply:
I think I'm getting ahead of myself here; for now we just want to use `opts.Limit` for the pipeline. Updated!